internal/mempool/mempool.go

```diff
@@ -1492,13 +1493,60 @@ func (mp *TxPool) maybeAcceptTransaction(tx *dcrutil.Tx, isNew, rateLimit,
 				log.Tracef("Transaction %v uses spent input %v and will be "+
 					"considered an orphan", txHash, hashCopy)
 			}
+
+			continue
+		}
+
+		// Check the fraud proof data.  If anything does not match, set the update
+		// fraud flag to true.  The fraud proof is not updated directly here since
+		// it requires copying the transaction, which we want to avoid unless
+		// necessary.
+		if int64(txIn.BlockHeight) != entry.BlockHeight() ||
+			txIn.BlockIndex != entry.BlockIndex() ||
+			txIn.ValueIn != entry.Amount() {
+
+			updateFraudProof = true
 		}
 	}
 
 	if len(missingParents) > 0 {
 		return missingParents, nil
 	}
 
+	// Update the fraud proof data on the transaction inputs as necessary.  The
+	// fraud proof data will ultimately get filled in properly when miners create
+	// a block template.  However, ensuring that it is filled in properly when
+	// entering the mempool is beneficial so that it is correct when relaying the
+	// transaction or returning it from an RPC API.
+	if updateFraudProof {
+		// Copy the transaction and swap the pointer.  This is to prevent races
+		// when modifying the fraud proof on the transaction inputs.
+		tx = dcrutil.NewTxDeepTxIns(tx)
```

Comment:
It's fine to use this as it is already implemented an exists, but just noting for myself when implementing the primitives bits that it would be ideal to have a variant that shallow copies the signature script to avoid more unneeded allocations.